cpu/cc2538/periph/uart.c

```diff
@@ -187,20 +206,76 @@ int uart_mode(uart_t uart, uart_data_bits_t data_bits, uart_parity_t parity,
 }
 #endif
 
+static inline void send_byte(uart_t uart, uint8_t byte)
+{
+    /* Block if the TX FIFO is full */
+    while (uart_config[uart].dev->cc2538_uart_fr.FRbits.TXFF) {}
+    uart_config[uart].dev->DR = byte;
+}
+
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
     assert(uart < UART_NUMOF);
-
-    cc2538_uart_t *u = uart_config[uart].dev;
-
-    /* Block if the TX FIFO is full */
+#ifdef MODULE_PERIPH_UART_NONBLOCKING
     for (size_t i = 0; i < len; i++) {
-        while (u->cc2538_uart_fr.FRbits.TXFF) {}
-        u->DR = data[i];
+        uart_config[uart].dev->cc2538_uart_im.IM |= TXMIS;
+        if (irq_is_in() || __get_PRIMASK()) {
+            /* if ring buffer is full free up a spot */
+            if (tsrb_full(&uart_tx_rb[uart])) {
+                send_byte(uart, tsrb_get_one(&uart_tx_rb[uart]));
+            }
+            /* if FIFO is full write to the buffer */
+            if (uart_config[uart].dev->cc2538_uart_fr.FRbits.TXFF) {
+                tsrb_add_one(&uart_tx_rb[uart], data[i]);
+            }
+            /* if tx FIFO is not full then add data to the FIFO */
+            else {
+                /* if there is already data in the tsrb buffer write that byte */
+                int byte = tsrb_get_one(&uart_tx_rb[uart]);
+                if (byte >= 0) {
+                    uart_config[uart].dev->DR = byte;
+                    tsrb_add_one(&uart_tx_rb[uart], data[i]);
+                }
+                /* If there is not data in the buffer directly write the
+                   current byte*/
+                else {
+                    uart_config[uart].dev->DR = data[i];
+                }
+            }
+        }
+        else {
+            /* If Tx FIFO is full then add to ring buffer */
+            if (uart_config[uart].dev->cc2538_uart_fr.FRbits.TXFF) {
+                while (tsrb_add_one(&uart_tx_rb[uart], data[i]) < 0) {}
+            }
+            /* If tx FIFO is not full then add data to the FIFO */
+            else {
+                /* if there is already data in the tsrb buffer write that byte */
+                /* need to disable IRQs in case FIFO gets empty enough
+                   to trigger an ISR */
+                unsigned state = irq_disable();
+                int byte = tsrb_get_one(&uart_tx_rb[uart]);
+                if (byte >= 0) {
+                    uart_config[uart].dev->DR = byte;
+                    irq_restore(state);
+                    while (tsrb_add_one(&uart_tx_rb[uart], data[i]) < 0) {}
+                }
+                /* If there is not data in the buffer directly write the
+                   current byte */
+                else {
+                    irq_restore(state);
+                    uart_config[uart].dev->DR = data[i];
+                }
+            }
```

Comment:
Why do we need this case?
Can't we just always write bytes to the TSRB when we are not in an interrupt / called with interrupts disabled?
`irq_handler_tx` already takes care of emptying the TSRB, so why duplicate it here?